numpy/core/shape_base.py

```diff
@@ -417,38 +419,99 @@ def format_index(index):
-    `result_ndim` is the greatest of the dimensions of the arrays in
-    `arrays` and the depth of the lists in `arrays` (see block docstring
-    for details).
+def _concatenate_shapes(shapes, axis):
+    """Given array shapes, return the resulting shape that would occur
+    after array concatenation.
+
+    concatenate(arrs, axis).shape == _concatenate_shapes([a.shape for a in arrs], axis)
     """
-    def atleast_nd(a, ndim):
-        # Ensures `a` has at least `ndim` dimensions by prepending
-        # ones to `a.shape` as necessary
-        return array(a, ndmin=ndim, copy=False, subok=True)
-
-    @recursive
-    def block_recursion(self, arrays, depth=0):
-        if depth < max_depth:
-            if len(arrays) == 0:
-                raise ValueError('Lists cannot be empty')
-            arrs = [self(arr, depth+1) for arr in arrays]
-            return _nx.concatenate(arrs, axis=-(max_depth-depth))
-        else:
-            # We've 'bottomed out' - arrays is either a scalar or an array
-            # type(arrays) is not list
-            return atleast_nd(arrays, result_ndim)
+    # Take a shape, any shape
+    first_shape = shapes[0]
+    for shape in shapes[1:]:
+        if (shape[:axis] != first_shape[:axis] or
+                shape[axis+1:] != first_shape[axis+1:]):
+            raise ValueError('Mismatched array shapes in block.')
+    shape_on_dim = sum(shape[axis] for shape in shapes)
+    return first_shape[:axis] + (shape_on_dim,) + first_shape[axis+1:]
+
+
+def _atleast_nd(a, ndim):
+    # Ensures `a` has at least `ndim` dimensions by prepending
+    # ones to `a.shape` as necessary
+    return array(a, ndmin=ndim, copy=False, subok=True)
+
+
+def _accumulate(values):
+    # Helper function because Python 2.7 doesn't have
+    # itertools.accumulate
+    value = 0
+    accumulated = []
+    for v in values:
+        value += v
+        accumulated.append(value)
 
-    return block_recursion(arrays)
+    return tuple(accumulated)
+
+
+def _block_info_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """Form required information to block the matricies using a flat iterator.
+
+    It returns a tuple containing
+
+    ``(shape, the_slices, the_arrays)``
+
+    * the resulting shape of the final array
+    * A flat iterable containing slices into the full array
+      required for assignment
+    * A flat iterable of arrays that need to be sliced in.
+
+    """
+    if depth < list_ndim:
+        concat_depth = result_ndim - list_ndim + depth
+        info = [_block_info_recursion(arr, list_ndim, result_ndim, depth + 1)
+                for arr in arrays]
+        shapes, slices, arrays = zip(*info)
+        # Compute the offset for each array
+        offsets = (0,) + tuple(_accumulate(shape[concat_depth]
+                                           for shape in shapes[:-1]))
+        # We contain lists that have the information from the lower dimensions
```

Comment:
Not sure what "we contain" refers to. There's no `self` argument to be an obvious candidate for "we"

Reply:
see rewording.